spot-contracts/contracts/vaults/RolloverVault.sol

```diff
@@ -0,0 +1,517 @@
+ *                 older tranche tokens (ones mature or approaching maturity) from perp.
+ *                 system through a rollover operation and earns an income in perp tokens.
+ *              2) `recover`: When executed, the vault redeems tranches for the underlying asset.
+ *                 NOTE: It performs both mature and immature redemption. Read more: https://bit.ly/3tuN6OC
+ *
+ *          At any time the vault will hold multiple ERC20 tokens, together referred to as the vault's "assets".
+ *          They can be a combination of the underlying asset, the earned asset and multiple tranches (deployed assets).
+ *
+ *          On redemption users burn their "notes" to receive a proportional slice of all the vault's assets.
+ *
+ *
+ */
+contract RolloverVault is
+    ERC20BurnableUpgradeable,
+    OwnableUpgradeable,
+    PausableUpgradeable,
+    ReentrancyGuardUpgradeable
+{
+    // data handling
+    using EnumerableSetUpgradeable for EnumerableSetUpgradeable.AddressSet;
+    using BondHelpers for IBondController;
+    using TrancheHelpers for ITranche;
+
+    // ERC20 operations
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+
+    // math
+    using MathUpgradeable for uint256;
+
+    //-------------------------------------------------------------------------
+    // Events
+
+    /// @notice Emits the vault asset's token balance that's recorded after a change.
+    /// @param token Address of token.
+    /// @param balance The recorded ERC-20 balance of the token.
+    event AssetSynced(IERC20Upgradeable token, uint256 balance);
+
+    //-------------------------------------------------------------------------
+    // Constants
+    uint8 public constant PERC_DECIMALS = 6;
+    uint256 public constant UNIT_PERC = 10**PERC_DECIMALS;
+    uint256 public constant HUNDRED_PERC = 100 * UNIT_PERC;
+
+    /// @dev Initial exchange rate between the underlying asset and notes.
+    uint256 private constant INITIAL_RATE = 10**6;
+
+    /// @dev Values should line up as is in the perp contract.
+    uint8 private constant PERP_PRICE_DECIMALS = 8;
+    uint256 private constant PERP_UNIT_PRICE = (10**PERP_PRICE_DECIMALS);
+
+    //--------------------------------------------------------------------------
+    // ASSETS
+    //
+    // The vault's assets are represented by a master list of ERC-20 tokens
+    //      => { [underlying] U _deployed U _earned }
+    //
+    // In the case of this vault, the "earned" assets are the perp tokens themselves.
+    // The reward (or yield) for performing rollovers is paid out in perp tokens.
+
+    /// @notice The ERC20 token that can be deposited into this vault.
+    IERC20Upgradeable public underlying;
+
+    /// @dev The set of the intermediate ERC-20 tokens when the underlying asset has been put to use.
+    ///      In the case of this vault, they represent the tranche tokens held before maturity.
+    EnumerableSetUpgradeable.AddressSet private _deployed;
```

Comment:
Any reason to make this private? We could remove the `deployedAt` function that's down below.

Reply:
I dont think, making `AddressSet` public doesn't automatically give us the public getter methods. They have to be implemented ..

Reply:
Ah of course, was thinking about built-in arrays.